eva/experimental/ray/executor/ray_remote.py

```diff
@@ -0,0 +1,61 @@
+# coding=utf-8
+# Copyright 2018-2022 EVA
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import os
+from typing import Callable, Dict, List
+
+import ray
+from ray.exceptions import RayTaskError
+from ray.util.queue import Queue
+
+from eva.executor.executor_utils import ExecutorError
+from eva.utils.logging_manager import logger
+
+
+class StageCompleteSignal:
+    pass
+
+
+@ray.remote(num_cpus=0)
+def ray_wait_and_alert(tasks: List[ray.ObjectRef], queue: Queue):
+    try:
+        ray.get(tasks)
+        queue.put(StageCompleteSignal)
+    except RayTaskError as e:
+        logger.error(e)
+        queue.put(ExecutorError(e.cause))
+
+
+@ray.remote(max_calls=1)
+def ray_parallel(
+    conf_dict: Dict[str, str],
+    executor: Callable,
+    input_queue: Queue,
+    output_queue: Queue,
+):
+    for k, v in conf_dict.items():
+        os.environ[k] = v
+
+    gen = executor(input_queue=input_queue)
+    for next_item in gen:
+        output_queue.put(next_item)
+
+
+@ray.remote(max_calls=1)
+def ray_pull(conf_dict: Dict[str, str], executor: Callable, input_queue: Queue):
+    for k, v in conf_dict.items():
+        os.environ[k] = v
+
+    for next_item in executor():
```

Comment:
Do you think we should do an explicit `executor.ray_exec()` instead of using the `__call__`? In most cases, `ray_exec()` or the current `__call__` are identical to `exec()`. Say, to support partition in seq scan, we can have a different `ray_exec()`.

Reply:
Yes, I think that is a good idea. @gaurav274

Reply:
Are you suggesting having 2 different execution logics for seq scan, `ray_exec` and `exec`?

Reply:
That is a possibility to support partition. We are not doing it in this pull request. I am suggesting we have an explicitly `ray_exec()` for readability. So later we modify the `ray_exec()` instead of `__call__` in case we want to have a different behavior when the executor is under ray.  By default, `ray_exec()` will `yield from exec()` in the abstract_executor like the current `__call__`

Reply:
I think it is a good idea. It could also be beneficial later if we want to extend to an operator that has more than one children.

Reply:
Can't we control this logic in `__call__`? I prefer keeping the logic in `ray_remote` clean.

Reply:
Something like this

```python
def ray_remote_func(executor):
    executor(remote=True)

Executor:
def exec(remote=False):
    if remote:
        # remote logic
    else:
        # local logic
```

I guess either way should work.

Reply:
Yes, I like this.

Reply:
I feel that this makes the exec more complicated and contains both the logic for parallel and non-parallel execution. The idea is to do it in `__call__`, but I think it's better to have an explicit name `ray_exec()` .

Reply:
Yes, my bad. I assumed it is in `__call__`